### PR TITLE
Log language client failures

### PR DESCRIPTION
**Pull request by akaroml**
Signed-off-by: Rome Li <rome.li@microsoft.com>

This is to record language client errors in the client log file.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -27,8 +27,47 @@ let lastStatus;
...
+class ClientErrorHandler implements ErrorHandler {
+	private restarts: number[];
+
+	constructor(private name: string) {
+		this.restarts = [];
+	}
+
+	public error(_error: Error, _message: Message, count: number): ErrorAction {
+		if (count && count <= 3) {
+			logger.error(`${this.name} server encounted error: ${_message}, ${_error && _error.toString()}`);
```
Comment (fbricon) [inline src/extension.ts]: encountered

File: src/extension.ts [hunk 2]
```diff
@@ -27,8 +27,47 @@ let lastStatus;
...
+			return ErrorAction.Continue;
+		}
+
+		logger.error(`${this.name} server encounted error and will shut down: ${_message}, ${_error && _error.toString()}`);
```
Comment (fbricon) [inline src/extension.ts]: encountered

**Comment by fbricon** [comment]
@akaroml please rebase against master and fix the tslint issues, or else the build will fail. See #1013

File: src/extension.ts [hunk 3]
```diff
@@ -27,8 +27,47 @@ let lastStatus;
...
+		this.restarts.push(Date.now());
+		if (this.restarts.length < 5) {
+			logger.error(`The ${this.name} server crashed and will restart.`);
+			return CloseAction.Restart;
+		} else {
+			let diff = this.restarts[this.restarts.length - 1] - this.restarts[0];
+			if (diff <= 3 * 60 * 1000) {
+				let message = `The ${this.name} server crashed 5 times in the last 3 minutes. The server will not be restarted.`;
+				logger.error(message);
+				window.showErrorMessage(message);
```
Comment (fbricon) [inline src/extension.ts]: Could we display buttons to open the client/server logs?

Comment (fbricon) [reply src/extension.ts]: And/or a button to open the troubleshooting page https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting

Comment (akaroml) [reply src/extension.ts]: That's exactly what I planned! The next PR will be showing both client and server logs side by side!

Comment (akaroml) [reply src/extension.ts]: #1016 was created. I'll work on a PR next week. Could you assign that issue to me?

**Comment by akaroml** [comment]
@fbricon can we get this merged first? I prefer a separate PR for #1016

**Comment by fbricon** [comment]
Thansk @akaroml !

**Pull request merged**